src/main/java/fi/metatavu/metamind/bot/OpenNLPSlotMatcher.java

```diff
@@ -0,0 +1,128 @@
+package fi.metatavu.metamind.bot;
+
+import java.net.URL;
+import java.util.HashMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.rabidgremlin.mutters.core.Context;
+import com.rabidgremlin.mutters.core.Intent;
+import com.rabidgremlin.mutters.core.Slot;
+import com.rabidgremlin.mutters.core.SlotMatch;
+import com.rabidgremlin.mutters.core.SlotMatcher;
+import com.rabidgremlin.mutters.core.Tokenizer;
+
+import opennlp.tools.namefind.NameFinderME;
+import opennlp.tools.namefind.TokenNameFinderModel;
+import opennlp.tools.util.Span;
+
+/**
+ * THIS IS A COPY OF ORIGINAL CLASS. REMOVE WHEN MUTTERS SUPPORTS LOADING SLOTS FROM URL
+ */
+
+/**
+ * Implements a SlotMatcher that uses OpenNLP's NER framework.
+ * 
+ * @author rabidgremlin
+ *
+ */
+public class OpenNLPSlotMatcher
+    implements SlotMatcher
+{
+  /** Logger. */
+  private Logger log = LoggerFactory.getLogger(OpenNLPSlotMatcher.class);
+
+  /** Map of NER models. */
+  private HashMap<String, TokenNameFinderModel> nerModels = new HashMap<String, TokenNameFinderModel>();
+
+  /** Map of slot models. These share the NER models. */
+  private HashMap<String, TokenNameFinderModel> slotModels = new HashMap<String, TokenNameFinderModel>();
+
+  /** The tokenizer to use. */
+  private Tokenizer tokenizer;
+
+  /**
+   * Constructor. Allows tokenizer to be supplied because NER can use case etc as cues, so may require different
+   * tokenizer than used for intent matching.
+   * 
+   * @param tokenizer The tokenizer to use on an utterance for NER.
+   */
+  public OpenNLPSlotMatcher(Tokenizer tokenizer)
+  {
+    this.tokenizer = tokenizer;
+  }
+
+  /**
+   * This set the NER model to use for a slot.
+   * 
+   * @param slotName The name of the slot. Should match the name of slots on intents added to the matcher.
+   * @param nerModel The file name of the NER model. This file must be on the classpath.
+   */
+  public void addSlotModel(String slotName, URL modelUrl)
+  {
+    TokenNameFinderModel tnfModel = nerModels.get(modelUrl.toExternalForm());
+    if (tnfModel == null)
+    {
+      try
+      {
+        tnfModel = new TokenNameFinderModel(modelUrl);
+      }
+      catch (Exception e)
+      {
+        throw new IllegalArgumentException("Unable to load NER model", e);
+      }
+    }
+
+    slotModels.put(slotName.toLowerCase(), tnfModel);
+  }
+
+  @Override
+  public HashMap<Slot, SlotMatch> match(Context context, Intent intent, String utterance)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') The return type of this method should be an interface such as "Map" rather than the implementation "HashMap". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1319)